index.js

```diff
@@ -18,149 +10,197 @@ var mkdirp = require('mkdirp');
+  const destpath = opts.increment ? incrementName(filepath) : filepath;
+
+  if (opts.overwrite === false && exists(filepath, destpath)) {
+    throw new Error('File already exists: ' + filepath);
+  }
+
+  mkdirSync(path.dirname(destpath), { recursive: true, ...options });
+  return fs.createWriteStream(destpath, opts);
 };
 
 /**
- * Uses `fs.createWriteStream` to write data to a file, replacing the
- * file if it already exists and creating any intermediate directories
- * if they don't already exist. Data can be a string or a buffer. Returns
- * a new [WriteStream](https://nodejs.org/api/fs.html#fs_class_fs_writestream)
- * object.
- *
- * ```js
- * var fs = require('fs');
- * var writeFile = require('write');
- * fs.createReadStream('README.md')
- *   .pipe(writeFile.stream('a/b/c/other-file.md'))
- *   .on('close', function() {
- *     // do stuff
- *   });
- * ```
- * @name .stream
- * @param {string|Buffer|integer} `filepath` filepath or file descriptor.
- * @param {object} `options` Options to pass to [mkdirp][] and [fs.createWriteStream][fs]{#fs_fs_createwritestream_path_options}
- * @return {Stream} Returns a new [WriteStream](https://nodejs.org/api/fs.html#fs_class_fs_writestream) object. (See [Writable Stream](https://nodejs.org/api/stream.html#stream_class_stream_writable)).
- * @api public
+ * Increment the filename if the file already exists and enabled by the user
  */
 
-writeFile.stream = function(filepath, options) {
-  mkdirp.sync(path.dirname(filepath), options);
-  return fs.createWriteStream(filepath, options);
+const incrementName = destpath => {
+  let file = { ...path.parse(destpath), path: destpath };
+  let name = file.name;
+  let n = 1;
+
+  while (fs.existsSync(file.path)) {
+    file.path = path.join(file.dir, `${name} (${++n})${file.ext}`);
+  }
+
+  return file.path;
+};
+
+/**
+ * Ensure newline at EOF if defined on options
+ */
+
+const ensureNewline = (data, options) => {
+  if (!options || options.newline !== true) return data;
+  if (typeof data !== 'string' && !isBuffer(data)) {
+    return data;
+  }
+
+  // only call `.toString()` on the last character. This way, if
+  // data is a buffer, we only need to call `.toString()` on
+  // the entire string if the condition is true.
+  if (String(data.slice(-1)) !== '\n') {
+    return data.toString() + '\n';
+  }
+
+  return data;
+};
+
+// if filepath !== destpath, that means the user has enabled
+// "increment", which has already checked the file system and
+// renamed the file to avoid conflicts, so we don't need to
+// check again.
+const exists = (filepath, destpath) => {
+  return filepath === destpath && fs.existsSync(filepath);
+};
+
+const mkdir = (dirname, options) => {
+  return new Promise(res => fs.mkdir(dirname, options, () => res()));
+};
+
+const mkdirSync = (dirname, options) => {
+  try {
+    fs.mkdirSync(dirname, options);
+  } catch (err) { /* do nothing */ }
+};
+
+const isBuffer = data => {
```

Comment:
Nice... I understand that my statement above about adding the newline to the buffer would only be possible if using the built-in `Buffer` class. But since you're doing this `isBuffer` check here, it seems intentional to leave `Buffer` out.

Reply:
Oh, damn. Good point. Hmm, we can either include the Buffer class, which - if I recall correctly - doesn't the native `fs` module need the buffer class anyway? Or we can just call `.toString()` if it's not a string already, before appending the `\n`. 

Thoughts?

Reply:
I think we leave `Buffer` in above, but we have an idea of what to do if people complain that they can't bundle `write` for some reason.